File: src/java/LeetCode/maxproduct_subarray/MaxProductSubArray_DP.java
package LeetCode.maxproduct_subarray;

public class MaxProductSubArray_DP {
    public static int maxProductDP(int[] nums) {
        if (nums == null || nums.length == 0) return 0;
        int maxSoFar = nums[0];
        int minSoFar = nums[0];
        int result = nums[0];
        for (int i = 1; i < nums.length; i++) {
            int num = nums[i];
            long a = (long) maxSoFar * num;
            long b = (long) minSoFar * num;
            int newMax = (int) Math.max(num, Math.max(a, b));
            int newMin = (int) Math.min(num, Math.min(a, b));
            maxSoFar = newMax;
            minSoFar = newMin;
            result = Math.max(result, maxSoFar);
        }
        return result;
    }

    public static void main(String args[]){
        int nums[] = new int[]{1,-2,-3,4};
        System.out.println("Max product subarray is : "+maxProductDP(nums));
    }
}

File: src/python/LeetCode150/maxproduct_subarray/maxproduct_subarray_dp.py
def maxProduct_dp(nums):
    if not nums:
        return 0
    max_so_far = min_so_far = result = nums[0]
    for num in nums[1:]:
        a = max_so_far * num
        b = min_so_far * num
        max_so_far = max(num, a, b)
        min_so_far = min(num, a, b)
        result = max(result, max_so_far)
    return result


# Example usage
if __name__ == "__main__":
    # Input: List of numbers
    nums = [7, 1, -5, 3, 6, -4]
    print("Maximum product sub array:", maxProduct_dp(nums))